File: packages/gavio-py/gavio/interceptors/pii/match.py
"""PiiMatch — a single detected PII entity within a span of text."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class PiiMatch:
    """One detected entity.

    ``start``/``end`` are half-open character offsets into the scanned text.
    ``replacement`` is the placeholder used in REDACT mode; ``value`` is the
    original text (never logged — used only for restore).
    """

    entity_type: str
    start: int
    end: int
    value: str
    confidence: float = 1.0
    replacement: str | None = None

    def __post_init__(self) -> None:
        if self.start < 0 or self.end < self.start:
            raise ValueError(
                f"Invalid PiiMatch span: start={self.start}, end={self.end}"
            )

    @property
    def length(self) -> int:
        return self.end - self.start

File: packages/gavio-js/src/interceptors/pii/match.ts
/** PiiMatch — a single detected PII entity within a span of text. */

/**
 * One detected entity.
 *
 * `start`/`end` are half-open character offsets into the scanned text.
 * `replacement` is the placeholder used in REDACT mode; `value` is the
 * original text (never logged — used only for restore).
 */
export interface PiiMatch {
  entityType: string
  start: number
  end: number
  value: string
  confidence: number
  /** e.g. '[EMAIL_1]'. */
  replacement: string
}

export function matchLength(m: PiiMatch): number {
  return m.end - m.start
}

export function makeMatch(init: {
  entityType: string
  start: number
  end: number
  value: string
  confidence?: number
  replacement: string
}): PiiMatch {
  if (init.start < 0 || init.end < init.start) {
    throw new Error(`Invalid PiiMatch span: start=${init.start}, end=${init.end}`)
  }
  return {
    entityType: init.entityType,
    start: init.start,
    end: init.end,
    value: init.value,
    confidence: init.confidence ?? 1.0,
    replacement: init.replacement,
  }
}
